File: src/main/java/com/jcloisterzone/game/capability/trait/FeatureCompletionBlocker.java
package com.jcloisterzone.game.capability.trait;

import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.state.GameState;

public interface FeatureCompletionBlocker {

	public boolean isFeatureCompletionBlocked(GameState state, FeaturePointer fp);

}

File: src/main/ts/com/jcloisterzone/game/capability/trait/FeatureCompletionBlocker.ts
import type { FeaturePointer } from "../../../board/pointer/FeaturePointer.js";
import type { GameState } from "../../state/GameState.js";

/** A capability that can veto the completion of a feature (Donkey, …).
 *  Mirrors Java {@code trait/FeatureCompletionBlocker}. */
export interface FeatureCompletionBlocker {
  isFeatureCompletionBlocked(state: GameState, fp: FeaturePointer): boolean;
}

/** Runtime mirror of Java's `instanceof FeatureCompletionBlocker` (duck-typed on
 *  the trait's single method). THE single definition. */
export function isInstanceOfFeatureCompletionBlocker(c: unknown): c is FeatureCompletionBlocker {
  return typeof (c as { isFeatureCompletionBlocked?: unknown }).isFeatureCompletionBlocked === "function";
}
